components/file_preview_modal/image_preview.tsx

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {useRef, useMemo, useEffect, useState} from 'react';
+import {clamp} from 'lodash';
+import classNames from 'classnames';
+
+import {getFilePreviewUrl, getFileDownloadUrl} from 'mattermost-redux/utils/file_utils';
+import {FileInfo} from '@mattermost/types/files';
+
+import {ZoomValue} from './file_preview_modal_image_controls/file_preview_modal_image_controls';
+import {LinkInfo} from './types';
+
+import './image_preview.scss';
+
+const HORIZONTAL_PADDING = 48;
+const VERTICAL_PADDING = 168;
+
+const SCROLL_SENSITIVITY = 0.003;
+const MAX_ZOOM = 5;
+
+let zoomExport: number;
+let minZoomExport: number;
+
+interface Props {
+    fileInfo: FileInfo & LinkInfo;
+    toolbarZoom: ZoomValue;
+    setToolbarZoom: (toolbarZoom: ZoomValue) => void;
+}
+
+// Utils
+const getWindowDimensions = () => {
+    const maxWidth = Math.max(document.documentElement.clientWidth || 0, window.innerWidth || 0) - HORIZONTAL_PADDING;
+    const maxHeight = Math.max(document.documentElement.clientHeight || 0, window.innerHeight || 0) - VERTICAL_PADDING;
+    return {maxWidth, maxHeight};
+};
+
+const fitCanvas = (width: number, height: number) => {
+    // Calculate maximum scale for canvas to fit in viewport
+    const {maxWidth, maxHeight} = getWindowDimensions();
+    const scaleX = maxWidth / width;
+    const scaleY = maxHeight / height;
+
+    return Math.round(Math.min(scaleX, scaleY) * 100) / 100;
+};
+
+export default function ImagePreview({fileInfo, toolbarZoom, setToolbarZoom}: Props) {
+    const background = useMemo(() => new Image(), []);
+    const {width, height} = background;
+    const containerScale = fitCanvas(width, height);
+
+    const zoom = useRef(0);
+    const minZoom = useRef(0);
+
+    minZoom.current = Math.min(containerScale, 1);
+    const maxCanvasZoom = containerScale;
+
+    let isFullscreen = {horizontal: false, vertical: false};
+    let cursorType = 'normal';
+
+    // Set the zoom given by the toolbar dropdown
+    switch (toolbarZoom) {
+    case 'A':
+        zoom.current = minZoom.current;
+        break;
+    case 'W':
+        zoom.current = getWindowDimensions().maxWidth / width;
```

Comment:
I would prefer a single call and destructuring it beforehand.

```
const {maxWidth, maxHeight} = getWindowDimensions();
```

Reply:
fixed